### ISSUE TITLE
Annotation to take priority in case of host conflict

### ISSUE THREAD

**Issue by benjamin-bergia** [nginxinc/kubernetes-ingress#1243]
**Is your feature request related to a problem? Please describe.**
I have nginx configured as my ingress on my cluster with cert-manager and http validation.
cert-manager creates ingresses during the certification creation and validation process. These ingresses hosts consistently collide with the created virtualservers hosts. 

**Describe the solution you'd like**
One simple solution would be to have an annotation that can be added to an ingress or VS to always take priority in case of host collision. After adding this annotation to the cert-manager ingress template, any temporary ingress created in the validation process would take priority over the application virtualserver. Once the certificate has been granted, cert-manager will remove the ingresses and the virtualserver will receive traffic again.

**Describe alternatives you've considered**
Looking at the way cert-manager generate its resources, I don't see any other simlle way. I would probably have to stop using VS and switch back to ingresses for all my apps, which at the moment is not possible.

**Additional context**
I do understand that it is not really an nginx ingress problem but it seems to be the simplest solution at the moment. Being able to create a new ingress or vs that will take priority over an existing one might also be useful during migrations or maintenance.

**Comment by anthonator** [comment]
I'm also interested in this specifically for the cert-manager use case. At the moment it doesn't seem possible to use cert-manager with virtualservers due to host collision.

**Comment by hostalp** [comment]
Did you try to use the `acme.cert-manager.io/http01-edit-in-place: "true"` annotation on the affected ingress resource? This solves the issue in my case, but it may not work with the virtualserver type.

The documentation for it is here https://cert-manager.io/docs/usage/ingress/#supported-annotations and it basically uses the existing ingress resource for the challenge instead of creating a new one (which would be conflicting with the existing one).

Another option is to use the `nginx.org/mergeable-ingress-type: master` annotation on the existing ingress resource and `nginx.org/mergeable-ingress-type: minion` on the temporary ones used to handle the challenge (https://github.com/nginxinc/kubernetes-ingress/tree/master/examples/mergeable-ingress-types), but I'm not sure if the latter can be achieved kinda transparently (without having to set it manually once the temporary ingress resource is created by the cert-manager).

However what I'd like to achieve is to be able to configure the ingress controller to add the `acme.cert-manager.io/http01-edit-in-place: "true"` annotation to all ingress resources by default so I don't have to take care of each ingress resource individually. That would be nice to have.

Some additional ingress controller option to merge the conflicting ingress resource configs would be of course great too. Just a simple logic ensuring that the most specific match wins or so.

**Comment by benjamin-bergia** [comment]
@hostalp my understanding is that `acme.cert-manager.io/http01-edit-in-place: "true"` would work if my application was using an ingress . But since I am using virtualservers only it shouldn't work. This is currently impacting all my services and none of them is able to renew their certificates.
In my opinion, the controller should throw errors by default and refuse to load a new ingress / vs that if it uses a server_name already in use, unless `mergeable-ingress-type` is specified. That would make the issue clearer.

**Comment by 030** [comment]
Addition of:
```
acme.cert-manager.io/http01-edit-in-place: "true"
```
did not work for us as well.

**Comment by hostalp** [comment]
@030
> Once the cert has been issued and the host is reverted to some-app.some-domain, the endpoint is valid.
And for which host is the new certificate issued? When you changed it to `some-app2.some-domain` it should be issued for this one so after going back to `some-app.some-domain` you should end up with a new one issued for your desired hostname.

One thing I noticed when adding the `acme.cert-manager.io/http01-edit-in-place: "true"` annotation was that I had to make sure a new certificate signing request is generated as the old one (generated when the config wasn't right) blocked any further operations for the same host. When you switched your config to other hostname and then back you may actually achieved that.
In short - if the new certificate is issued for `some-app.some-domain` (not for `some-app2.some-domain`) then you're likely fine.

By default the certificate is going to be renewed 30 days before the expiration of the current one so you'll have enough time to check whether it really works. Or create some test ingress and retry the same thing with that.

@benjamin-bergia
Yes I kind of expected it wouldn't work with the virtualserver resource. The Nginx docs mention that there's simply no merging support for virtualservers at all: https://docs.nginx.com/nginx-ingress-controller/configuration/handling-host-collisions/#merging-configuration-for-the-same-host
It looks like cert-manager should ideally support virtualservers to handle this correctly.

**Comment by benjamin-bergia** [comment]
@hostalp yeah agreed, but I don't think it's doable on cert-manager side to support all types of ingress resources under the sun. Maybe supporting templates for any kind of resources might be a solution on cert-manager side, in any case I would still be interested, on nginx side, by being able to define two resources (ingress or virtualserver) with the same host and manually set (via an annotation) which one is taking priority over the other. On top of solving the cert-manager case it will help with maintenance pages, and other situations where you temporarily want to enable/disable some services without rewriting your whole ingress.

@030 Unfortunately, I didn't find any solution at the moment. If you use gitops and all your conf is on a repo it makes it even worst since you can't manually switch domains like in your example.

**Comment by benjamin-bergia** [comment]
Actually, I don't think letting cert-manager modify my virtualServer or ingress in place is viable. I use argocd to manage my application. It means that any change to a resource that is not also changed in my repo will be rolled back. The only solution with gitops CD systems is for cert-manager to create a new, un-managed, ingress.

**Comment by 030** [comment]
@hostalp The `acme.cert-manager.io/http01-edit-in-place: "true"` seems to solve the issue when a new ingress is deployed.

**Comment by benjamin-bergia** [comment]
So after some more work and few trial and errors, here is the **only** solution that allows to tie together nginx-ingress, cert-manager and argocd when using http01 challenge.

1. Configure cert-manager to create mergeable ingresses by adding the following to your issuer or clusterIssuer:
```yaml
            ingressTemplate:
              metadata:
                annotations:
                  nginx.org/mergeable-ingress-type: minion
```
2. Create your ingresses with the following annotations:
```yaml
annotations:
    kubernetes.io/ingress.class: nginx
    cert-manager.io/cluster-issuer: my-issuer
    cert-manager.io/issue-temporary-certificate: 'true'
    nginx.org/mergeable-ingress-type: master
```
3. Move all your rules out of your master ingress and into your minion as mentioned in the doc.


A complete, minimum, example would be:
```yaml
---
apiVersion: cert-manager.io/v1alpha2
kind: ClusterIssuer
metadata:
  name: acme-staging
spec:
  acme:
    email: ***
    server: https://acme-staging-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
      name: acme-staging-key
    solvers:
      - http01:
          ingress:
            class: nginx
            serviceType: ClusterIP
            ingressTemplate:
              metadata:
                annotations:
                  nginx.org/mergeable-ingress-type: minion

---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: my-app
  annotations:
    kubernetes.io/ingress.class: nginx
    ingress.kubernetes.io/ssl-redirect: 'true'
    nginx.org/mergeable-ingress-type: master
    cert-manager.io/cluster-issuer: acme-staging
    cert-manager.io/issue-temporary-certificate: 'true'
spec:
  tls:
    - hosts: [my.app.com]
      secretName: my-app-cert
  rules:
    - host: my.app.com

---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: my-app-frontend
  annotations:
    kubernetes.io/ingress.class: nginx
    nginx.org/mergeable-ingress-type: minion
spec:
  rules:
    - host: my.app.com
      http:
        paths:
            backend:
              serviceName: my-app
              servicePort: 80

```

**Comment by benjamin-bergia** [comment]
I realized that the temporary certificate annotation cannot be added to an ingress but a cert only. It means that I can't use the automated generation of certificates and had to create certificates for all my ingresses. I updated my previous example.

**Comment by svvac** [comment]
I had to resort to the workaround by @benjamin-bergia. Also note that sine mergeable ingresses only support a single host, you need to duplicate the ingress stack (master + minions) for each supported host, even if they share the same certificate.

The saddest part is that you can't benefit from the features of the v2 API (Policies, OIDC, etc.) if you want to provision using ACME HTTP01.

**Comment by timvol** [comment]
I know the issue has been closed, but I don't wanted to create a new issue ticket for this...

@benjamin-bergia I added all the annotations to my issuer and my ingress, but it finally says "Ingress master is invalid or doesn't exist" and "spec.rules[0].http.paths: Too many: 1: must have at most 0 items". My configuration as follows:

```yaml
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: letsencrypt-prod
spec:
  acme:
    email: mymail@something.tld
    server: https://acme-staging-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
        name: example-issuer-account-key
    solvers:
      - http01:
          ingress:
            class: nginx
            serviceType: ClusterIP
            ingressTemplate:
                metadata:
                    annotations:
                        nginx.org/mergeable-ingress-type: minion
```

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example-ingress
  annotations:
    kubernetes.io/ingress.class: nginx
    cert-manager.io/issuer: letsencrypt-prod
    nginx.org/mergeable-ingress-type: master
spec:
  rules:
  - host: whoami.something.tld
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: whoami
            port:
              number: 80
  tls:
  - hosts:
    - whoami.something.tld
    secretName: acme-crt-secret
```

**Comment by timvol** [comment]
Thank you @svvac Now I'm able to create mergeable ingresses. But unfortunately I cannot create a certificate since the master doesn't expose port 80. It exposes just port 443 and when I try to access the page, it tells me "ERR_SSL_VERSION_OR_CIPHER_MISMATCH". I assume the problem here is that the certificate I specified in the master isn't valid/generated because the Let's Encrypt challenge fails with "Waiting for HTTP-01 challenge propagation: failed to perform self check GET request 'http://sub.something.com/.well-known/acme-challenge/5VN5mBmX6...': Get "https://sub.something.com:443/.well-known/acme-challenge/5VN5mBmX6...": remote error: tls: handshake failure"

My master looks like:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: my-app-ingress-master1
  annotations:
    kubernetes.io/ingress.class: nginx
    nginx.org/redirect-to-https: "false"
    nginx.org/mergeable-ingress-type: "master"
    cert-manager.io/cluster-issuer: letsencrypt
spec:
  tls:
  - hosts:
    - sub.something.com
    secretName: acme-crt-secret
  rules:
    - host: sub.something.com
```

I guess it's because Let's Encrypt tries to access the http-version on port 80 but it's getting redirected to port 443 (where I'm missing the tls certificate I need to generate using Let's Encrypt). I already tried to set `nginx.org/redirect-to-https: "false"` but it didn't help. So, any idea how I can fix this?

**Comment by benjamin-bergia** [comment]
@timvol Unfortunately I never managed to get it to work properly and since then I have moved to a different ingress controller.
like @svvac mentioned, you need an empty master ingress and one minion for your actual rules. Your master ingress should have no rules.

**Comment by timvol** [comment]
@benjamin-bergia Which ingress controller are you using now?

**Comment by benjamin-bergia** [comment]
@timvol sorry but I don't think it would be fair to advertise other controllers here. I pretty much went through the [list](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/) and tested them one by one. It should be possible to run this controller with cert-manager and mergeable ingresses. Unfortunately this didn't play well with my CD system.

Like @svvac  mentioned there is a bit of a catch since your master ingress needs to be empty (without any rule) and cannot contain some annotations.

**Comment by svvac** [comment]
@timvol You might need the following annotation to issue a temporary self-signed cert until LetsEncrypt gives you one, so that you can actually connect using TLS : 

```
cert-manager.io/issue-temporary-certificate: 'true'
```

I have this set directly on the `Certificate` object, but it appears to [also be recognized on the `Ingress`](https://cert-manager.io/docs/usage/ingress/#supported-annotations).

Personally I've since moved to DNS01 challenges because HTTP01 is such a pain with this ingress controller (and not doable so far if you want to use `VirtualServer`s and the likes).

**Comment by timvol** [comment]
Some days ago I also switched to another ingress controller but today I ran into other problems and finally switched back to nginx and I was finally able to solve the issue in nginx.
One issue was that I got the error, that the master ingress does not exist. The reason was the order of the deployment. Since I'm using ArgoCD to manage my applications, everything was deployed at the same time. Sometimes the master ingress was created before the "minion"-ingress and sometimes the "minions" were created before the master. Since my application is a Helm chart, I was able to use the `helm.sh/hook-weight: "-5"` annotation on my master ingress to ensure that it's getting created before the minions.
Afterwards I got the problem mentioned some comments above by myself (the `remote error: tls: handshake failure`). And in this case, the comment of @svvac helped me. Unfortunately it didn't worked at first because I've added it to the master ingress. Maybe I've done something wrong but I was finally get it working after applying the `cert-manager.io/issue-temporary-certificate: "true"` to my certificate. Now, everything works as expected.Thanks @svvac and @benjamin-bergia
